lib/class-wp-rest-image-editor-controller.php

```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
-		$media_url       = wp_get_attachment_image_url( $media_id, 'original' );
-
-		if ( ! $attachment_info || ! $media_url ) {
+		if ( ! $image_meta || ! $image_file ) {
 			return new WP_Error( 'rest_unknown_attachment', __( 'Unable to get meta information for file.', 'gutenberg' ), array( 'status' => 404 ) );
 		}
 
-		$meta = array( 'original_name' => basename( $media_url ) );
+		$image_editor = wp_get_image_editor( $image_file );
 
-		if ( isset( $attachment_info['richimage'] ) ) {
-			$meta = array_merge( $meta, $attachment_info['richimage'] );
+		if ( is_wp_error( $image_editor ) ) {
+			// This image cannot be edited.
+			return new WP_Error( 'rest_unknown_image_file_type', __( 'Unable to edit this image.', 'gutenberg' ), array( 'status' => 500 ) );
 		}
 
-		// Try and load the image itself.
-		$image_path = get_attached_file( $media_id );
-		if ( empty( $image_path ) ) {
-			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
-		}
+		if ( isset( $params['rotation'] ) ) { // TODO: use `rotate` instead?
+			// CW / CCW
+			$result = $image_editor->rotate( 0 - $params['rotation'] );
 
-		$image_editor = wp_get_image_editor( $image_path );
-		if ( ! $image_editor->load() ) {
-			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+			// TODO: return the "real" error message?
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_rotation_failed', __( 'Unable to rotate this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		if ( isset( $params['rotation'] ) ) {
-			$image_editor->rotate( 0 - $params['rotation'] );
+		if ( isset( $params['x'], $params['y'], $params['width'], $params['height'] ) ) {
+			$size = $image_editor->get_size();
+
+			// Crop
+			$crop_x = round( ( $size['width'] * floatval( $params['x'] ) ) / 100.0 );
+			$crop_y = round( ( $size['height'] * floatval( $params['y'] ) ) / 100.0 );
+			$width  = round( ( $size['width'] * floatval( $params['width'] ) ) / 100.0 );
+			$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
+
+			$result = $image_editor->crop( $crop_x, $crop_y, $width, $height );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_crop_failed', __( 'Unable to crop this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		$size = $image_editor->get_size();
+		// Calculate the file name.
+		$image_ext  = pathinfo( $image_file, PATHINFO_EXTENSION );
+		$image_name = wp_basename( $image_file, ".{$image_ext}" );
+
+		// Do not append multiple `-edited` to the file name.
+		// The user may be editing a previously edited image.
+		if ( preg_match( '/-edited(-\d+)?$/', $image_name ) ) {
+			// Remove any `-1`, `-2`, etc. `wp_unique_filename()` will add the proper number.
+			$image_name = preg_replace( '/-edited(-\d+)?$/', '-edited', $image_name );
```

Comment:
Can't we replace immediately? Do we need to match first?

Reply:
Need it for the `else` to append`-edited`.